### PR TITLE
BUFR2IODA python API converters for marine monthly in situ BUFR

### LINKED ISSUE #711: in situ BUFR IODA converters using BUFR2IODA python API
The IODA converters for in situ BUFR (monthly and daily tanks) using BUFR2IODA python API are provided for:
**monthly**
- [x] ARGO
- [x] GLIDER 
- [x] BATHY
- [x] TESAC
- [x] TESAC MAMMALS
- [x] XBTCTD
- [x] ALTKOB
- [ ] ~~AXBT~~ **No data available**
**daily**
- [ ] DBUOY
- [ ] DBUOYB
- [ ] MBUOY
- [ ] MBUOYB
- [ ] SALDRN

### PR DESCRIPTION
**Pull request by ShastriPaturi**
This PR adds bufr2ioda python API converters for marine monthly in situ data based on the work done during the "End-to-End Code Sprint" by the atmosphere group.

partially resolves #711

### CODE REVIEW

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py [hunk 1]
```diff
@@ -20,15 +20,13 @@
 from pyioda import ioda_obs_space as ioda_ospace
```
Comment (guillaumevernieres) [inline ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py]: Minor issue: I don't know anything about that script, but your text editor is poorly setup. You're adding spaces a the end of a few lines below.

**Review by guillaumevernieres — COMMENTED** [review]
Thanks @ShastriPaturi !
1 - Your choice , but I thought @rmclaren said we could use `yaml` instead of `jason`.
2 - I'm not reviewing the code, but it looks like you're having "extra space" issues with your text editor. 
3 - code norm's failing
Are these converters tested as part of the ci?

**Comment by ShastriPaturi** [comment]
> Thanks @ShastriPaturi ! 1 - Your choice , but I thought @rmclaren said we could use `yaml` instead of `jason`. 2 - I'm not reviewing the code, but it looks like you're having "extra space" issues with your text editor. 3 - code norm's failing Are these converters tested as part of the ci?
1 - Your choice , but I thought @rmclaren said we could use yaml instead of jason
     Following what was done in the [End-to-End] Code sprint
2 - I'm not reviewing the code, but it looks like you're having "extra space" issues with your text editor.
     fixing
3 - code norm's failing
     fixing
Are these converters tested as part of the ci?
   Not that I know of

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py (hunk 1 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py]: Fixed.

File: ush/ioda/bufr2ioda/bufr2ioda_bathythermal_profiles.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    # PreQC
+    obsspace.create_var('PreQC/waterTemperature', dtype=PreQC.dtype, fillval=PreQC.fill_value) \
+        .write_attr('long_name', 'PreQC') \
+        .write_data(PreQC)
+
+    # ObsValue
+    obsspace.create_var('ObsValue/waterTemperature', dtype=temp.dtype, fillval=temp.fill_value) \
+        .write_attr('units', 'degC') \
+        .write_attr('valid_range', np.array([-10.0, 50.0], dtype=np.float32)) \
+        .write_attr('long_name', 'Temperature below surface') \
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_bathythermal_profiles.py]: Might suggest calling the long_name simply "water Temperature" unless it is guaranteed that every measurement has depth to it.

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py [hunk 3]
```diff
@@ -246,10 +248,10 @@ def bufr_to_ioda(config, logger):
         .write_data(stationID)
 
     # Depth
-    obsspace.create_var('MetaData/depth', dtype=depth.dtype, fillval=depth.fill_value) \
+    obsspace.create_var('MetaData/depth', dtype=pressure.dtype, fillval=pressure.fill_value) \
         .write_attr('units', 'm') \
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py]: Can you explain how/why depth is a pressure here? If it's a pressure, should it be 'Pa' ?

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_glider_profiles.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        .write_data(stationID)
+
+    # Depth
+    obsspace.create_var('MetaData/depth', dtype=pressure.dtype, fillval=pressure.fill_value) \
+        .write_attr('units', 'm') \
+        .write_attr('long_name', 'Water depth') \
+        .write_data(pressure)
+
+    # Sequence Number
+    obsspace.create_var('MetaData/sequenceNumber', dtype=PreQC.dtype, fillval=PreQC.fill_value) \
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_subpfl_glider_profiles.py:257]: Confirming that you want this to have the same dtype as PreQC and not sequenceNumber specific?

File: ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
+#!/usr/bin/env python3
+# (C) Copyright 2023 NOAA/NWS/NCEP/EMC
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py]: I'll only comment once but consider changing these to 2024

File: ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    obsspace.create_var('ObsError/salinity', dtype=ObsError_saln.dtype, fillval=ObsError_saln.fill_value) \
+        .write_attr('units', 'psu') \
+        .write_attr('long_name', 'ObsError') \
+        .write_data(ObsError_saln)
+
+    # ObsValue
+    obsspace.create_var('ObsValue/waterTemperature', dtype=temp.dtype, fillval=temp.fill_value) \
+        .write_attr('units', 'degC') \
+        .write_attr('valid_range', np.array([-10.0, 50.0], dtype=np.float32)) \
+        .write_attr('long_name', 'Temperature below surface') \
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py]: Same as before. You might want to keep these make these "below surface" be "water Temperature" unless it's guaranteed that all of the measurements have depth.

File: ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        .write_data(temp)
+
+    obsspace.create_var('ObsValue/salinity', dtype=saln.dtype, fillval=saln.fill_value) \
+        .write_attr('units', 'psu') \
+        .write_attr('valid_range', np.array([0.0, 45.0], dtype=np.float32)) \
+        .write_attr('long_name', 'Salinity below surface') \
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py]: See above.

File: ush/ioda/bufr2ioda/bufr2ioda_altkob_surface.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+from pyiodaconv import bufr
+from collections import namedtuple
+from pyioda import ioda_obs_space as ioda_ospace
+from wxflow import Logger
+
+
+def Compute_sequenceNumber(lon):
+
+    lon_u, seqNum = np.unique(lon, return_inverse=True)
+    seqNum = seqNum.astype(np.int64)
```
Comment (nicholasesposito) [inline ush/ioda/bufr2ioda/bufr2ioda_altkob_surface.py]: Usually we only want time variables to be np.int64. Recommending you change to np.int32 unless @emilyhcliu thinks it's fine.

File: ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py (hunk 6 above)
Comment (nicholasesposito) [reply ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py]: I'll keep this as the main comment to stop this from getting messy, but keep this in mind for the other files too.

**Review by nicholasesposito — COMMENTED** [review]
No major issues besides int32 vs int64 for sequenceNumber, though there are some things to consider/recommendations that I have.

**Comment by ShastriPaturi** [comment]
> No major issues besides int32 vs int64 for sequenceNumber, though there are some things to consider/recommendations that I have.

I agree and addressed.

File: ush/ioda/bufr2ioda/bufr2ioda_bathythermal_profiles.py (hunk 2 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_bathythermal_profiles.py]: Accepted. Thanks

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py (hunk 3 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_subpfl_argo_profiles.py]: Lapse on my end. Fixed.

File: ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py (hunk 5 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_tesac_mammals_profiles.py]: Done.

File: ush/ioda/bufr2ioda/bufr2ioda_altkob_surface.py (hunk 8 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_altkob_surface.py]: Agreed and updated.

File: ush/ioda/bufr2ioda/bufr2ioda_subpfl_glider_profiles.py (hunk 4 above)
Comment (ShastriPaturi) [reply ush/ioda/bufr2ioda/bufr2ioda_subpfl_glider_profiles.py:257]: yes.

**Review by nicholasesposito — APPROVED** [review]
Thanks for making the changes. Looks good!

**Comment by ShastriPaturi** [comment]
> Thanks for making the changes. Looks good!

Thanks a lot @nicholasesposito. Appreciate your quick response.

**Pull request merged**